database/src/ops/block.rs

```diff
@@ -1,89 +1,327 @@
+/// // TODO
+/// ```
+#[doc = doc_error!()]
+#[allow(clippy::too_many_lines)]
+// no inline, too big.
+pub fn add_block(
+    tables: &mut impl TablesMut,
+    block: VerifiedBlockInformation,
+) -> Result<(), RuntimeError> {
+    let VerifiedBlockInformation {
+        block,
+        txs,
+        block_hash,
+        pow_hash,
+        height,
+        generated_coins,
+        weight,
+        long_term_weight,
+        cumulative_difficulty,
+        block_blob,
+    } = block;
 
-/// TODO
-pub fn get_block_hash() {
-    todo!()
-}
+    let cumulative_rct_outs = crate::ops::output::get_rct_num_outputs(tables.rct_outputs_mut())?;
 
-/// TODO
-pub fn get_block_height() {
-    todo!()
-}
+    // Block Info.
+    tables.block_infos_mut().put(
+        &height,
+        &BlockInfo {
+            timestamp: block.header.timestamp,
+            total_generated_coins: generated_coins,
+            cumulative_difficulty,
+            block_hash,
+            cumulative_rct_outs, // TODO: should this be a dummy value?
+            // INVARIANT: #[cfg] @ lib.rs asserts `usize == u64`
+            weight: weight as u64,
+            long_term_weight: long_term_weight as u64,
+        },
+    )?;
 
-/// TODO
-pub fn get_block_weight() {
-    todo!()
-}
+    // Block blobs.
+    tables
+        .block_blobs_mut()
+        .put(&height, &StorableVec(block_blob))?;
 
-/// TODO
-pub fn get_block_already_generated_coins() {
-    todo!()
-}
+    // Block heights.
+    tables.block_heights_mut().put(&block_hash, &height)?;
 
-/// TODO
-pub fn get_block_long_term_weight() {
-    todo!()
-}
+    // Transaction & Outputs.
+    for tx in txs {
+        let tx: &Transaction = &tx.tx;
+        add_tx(tx, tables)?;
 
-/// TODO
-pub fn get_block_timestamp() {
-    todo!()
-}
+        // ^
+        // Everything above is adding tx data.
+        // Everything below is for adding input/output data.
+        // v
 
-/// TODO
-pub fn get_block_cumulative_rct_outputs() {
-    todo!()
+        // Is this a miner transaction?
+        // Which table we add the output data to depends on this.
+        // <https://github.com/monero-project/monero/blob/eac1b86bb2818ac552457380c9dd421fb8935e5b/src/blockchain_db/blockchain_db.cpp#L212-L216>
+        let mut miner_tx = false;
+
+        for inputs in &tx.prefix.inputs {
+            match inputs {
+                // Key images.
+                Input::ToKey { key_image, .. } => {
+                    add_key_image(tables.key_images_mut(), key_image.compress().as_bytes())?;
+                }
+                // This is a miner transaction, set it for later use.
+                Input::Gen(_) => miner_tx = true,
+            }
+        }
+
+        // Output bit flags.
+        // Set to a non-zero bit value if the unlock time is non-zero.
+        let output_flags = match tx.prefix.timelock {
+            Timelock::None => 0x0000_0000,
+            Timelock::Block(_) | Timelock::Time(_) => 0x0000_0001,
+        };
```

Comment:
Is this the correct bit value?

Reply:
Yeah, might be a good idea to use `bitflags`

Reply:
> Inside our cuprate_database::types::Output { .., output_flags: u32 }? Or just as a type for external use? It'll mess with bytemuck if we include the type in the struct I think.

we can do:

```rust
#[derive(Pod, etc)]
#[repr(transparent)]
pub struct OutputFlags(u32;

bitflags! {
    impl OutputFlags: u32 {
        const TIME_LOCKED = 0b0000_0001;
     }
}
```

Which should work right?

> Height's are u64 so add_block() doesn't work unless you wanted to cast since 500_000_000 < u32::MAX?

Yeah just do a cast. Do `try_into` though so we panic if we somehow get a higher height when adding a block (should never happen but better than a silent corrupted DB).